Kuadrant/limitador-operator
# feat: immediate limits sync to pod via annotation

# Description
Part of https://github.com/Kuadrant/kuadrant-operator/issues/414

This PR adds a feature to annotate Limitador pods with the limits config map resource version. This change ensures that modifications to the limits will promptly trigger a reload or synchronization of the limit configuration to Limitador pods. Consequently, it ensures that pod limits are always immediately up to date.

This enhancement will enable us to utilize the Limitador CR `Ready` condition as a way of naively assuming that limits are enforced.

Other additions include
* Updating kind to v0.22.0
* Refactor integration tests to use context to timeout tests and algin primarily use Gomega for assertions

# Verification
The following have already been verified in the included integration test but if you want to verify manually:
* Checkout this branch
* Create local cluster
```
 make local-setup
```
* Watch for limitador pod annotations
```
watch "kubectl get pods -o yaml | yq '.items[].metadata.annotations'"
```
* Deploy limitador CR
```
kubectl apply -f -<<EOF                                          
apiVersion: limitador.kuadrant.io/v1alpha1
kind: Limitador
metadata:
  name: limitador-sample
spec:
  replicas: 2
  limits:
    - conditions: ["get_toy == 'yes'"]
      max_value: 10
      namespace: toystore-app
      seconds: 30
      variables: []
EOF
```
* Verify annotation is added to the pods
* Watch limits in a limitador pod
```
watch "kubectl exec $(kubectl get pods -l app=limitador -o yaml | yq '.items[0].metadata.name') -c limitador -- cat /home/limitador/etc/limitador-config.yaml"
```
* Update limits
```
kubectl apply -f -<<EOF                                          
apiVersion: limitador.kuadrant.io/v1alpha1
kind: Limitador
metadata:
  name: limitador-sample
spec:
  replicas: 2
  limits:
    - conditions: ["get_toy == 'yes'"]
      max_value: 20
      namespace: toystore-app
      seconds: 60
      variables: []
EOF
```
* Verify annotation of config map resource version is updated
* Verify config map in pod is immediately updated / synced to new value once annotation is updated
